crates/codegen/runtime/cargo/src/runtime/diagnostic.rs

```diff
@@ -0,0 +1,65 @@
+use std::borrow::Cow;
+use std::error::Error;
+
+use crate::text_index::TextRange;
+
+#[repr(u8)]
+pub enum Severity {
+    Error = 1,
+    Warning = 2,
+    Information = 3,
+    Hint = 4,
+}
+
+pub trait Diagnostic: Error {
+    fn range(&self) -> TextRange;
```

Comment:
I've decided to follow the design of the LSP and have diagnostics be local to a given resource.

In the future, we could have something like `DiagnosticWithSpan` that has the ability to fetch a full span (i.e. source + range).

Reply:
Sounds good. But I think @AntonyBlakey is already working on adding column tracking? so this comment might be obsolete very soon.

Reply:
Additionally, I think we are using `text_range()` everywhere. I wonder if there is a reason this is just `range()`?

Reply:
LSP uses `range`, so I decided to mirror that API here.

Alternatively, we could use `text_range` here and, instead, provide a separate `LspDiagnostic` trait that mirrors LSP 100% that the types implementing `Diagnostic` might be convertible to.

Let me know what's your preference.

Reply:
Not sure mirroring the exact field names of the LSP is a goal here?
I think the goal is to provide diagnostics that users can convert to LSP-like messages. This means things like severity, a string description, line/column range, etc...
But there are other things that we will never add, like related information, tags, "source" fields, etc...
This way, any user can easily convert our diagnostics into other common error reporting APIs. But for the exact LSP API, I think `hardhat-vscode` is going to be the only case for this for the foreseable future. This is why I see having a consistent API/terminology accross our entire API is more important.

What do you think?